### PR TITLE
[modifier refactor] Add constant pruning tests 

### PR DESCRIPTION
**Pull request by rahul-tuli**
This PR serves as the base branch for all `ConstantPruningModifier` tests:

- [X] End to End tests
- [X] Serialization Bugfix #1753 
- [X] Serialization Test #1755 
- [X] Unit Tests #1756 
- [X] Mask De-registration Bugfix #1773 
- [X] Session reset for all tests #1790

### CODE REVIEW

File: tests/sparseml/core/recipe/test_recipe.py [hunk 1]
```diff
@@ -14,30 +14,78 @@
...
+@pytest.mark.parametrize("recipe_str", _valid_recipes())
+def test_serialization(recipe_str):
+    recipe_instance = Recipe.create_instance(recipe_str)
+    serialized_recipe_str = recipe_instance.yaml()
+    recipe_from_serialized = Recipe.create_instance(serialized_recipe_str)
+
+    expected_dict = recipe_instance.dict()
+    actual_dict = recipe_from_serialized.dict()
+
+    assert expected_dict == actual_dict
```
Comment (Satrat) [inline tests/sparseml/core/recipe/test_recipe.py:91]: could we test initializing from a dict and from explicitly defining the arguments as well? I believe a lot of the other serialization tests do this

File: tests/sparseml/modifiers/pruning/constant/test_pytorch.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    # move imports inside so that pytorch is not required unless
+    # running this test
+
+    from sparseml.core import State
+    from sparseml.core.event import Event, EventType
+    from sparseml.core.framework import Framework
+    from sparseml.modifiers.pruning.constant.pytorch import (
+        ConstantPruningModifierPyTorch,
+    )
+    from sparseml.pytorch.utils import tensor_sparsity
```
Comment (Satrat) [inline tests/sparseml/modifiers/pruning/constant/test_pytorch.py:88]: Since the Makefile is set to ignore this file except when we run the pytorch tests can't we move these to the top of the file?

File: Makefile [hunk 3]
```diff
@@ -25,7 +25,7 @@ ifneq ($(findstring onnx,$(TARGETS)),onnx)
     PYTEST_ARGS := $(PYTEST_ARGS) --ignore tests/sparseml/onnx
 endif
 ifneq ($(findstring pytorch,$(TARGETS)),pytorch)
-    PYTEST_ARGS := $(PYTEST_ARGS) --ignore tests/sparseml/pytorch
+    PYTEST_ARGS := $(PYTEST_ARGS) --ignore tests/sparseml/pytorch --ignore tests/sparseml/modifiers
```
Comment (Satrat) [inline Makefile:28]: We should think about a better way to separate out the pytorch tests, because eventually we will have other frameworks in this folder. This will also skip any framework-independent base tests we write

File: tests/sparseml/modifiers/pruning/constant/test_pytorch.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+        if "weight" in name
+    }
+    assert manipulated_sparsities != expected_sparsities, "Sparsity manipulation failed"
+
+    # apply modifier
+
+    modifier.on_update(state, event=Event(type_=EventType.OPTIM_PRE_STEP))
+    modifier.on_update(state, event=Event(type_=EventType.OPTIM_POST_STEP))
+    modifier.finalize(state)
+    modifier.on_end(state, None)
```
Comment (Satrat) [inline tests/sparseml/modifiers/pruning/constant/test_pytorch.py:127]: these should be called in the opposite order

File: tests/sparseml/modifiers/pruning/constant/test_pytorch.py [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+        optimizer=optimizer(model.parameters(), lr=0.1),
+        start=0,
+    )
+    modifier = ConstantPruningModifierPyTorch(
+        targets="__ALL_PRUNABLE__",
+        start=0,
+        end=1,
+        update=0.5,
+    )
+    modifier.initialize(state)
```
Comment (Satrat) [inline tests/sparseml/modifiers/pruning/constant/test_pytorch.py:110]: Could we add a check after this call to make sure the masks have been added to the model and that they match up with the 0 locations?

File: tests/sparseml/modifiers/pruning/constant/test_pytorch.py [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+        for name, param in model.named_parameters()
+        if "weight" in name
+    }
+    assert manipulated_sparsities != expected_sparsities, "Sparsity manipulation failed"
+
+    # apply modifier
+
+    modifier.on_update(state, event=Event(type_=EventType.OPTIM_PRE_STEP))
+    modifier.on_update(state, event=Event(type_=EventType.OPTIM_POST_STEP))
+    modifier.finalize(state)
```
Comment (Satrat) [inline tests/sparseml/modifiers/pruning/constant/test_pytorch.py:126]: Lets test that the masks were actually removed after this call

File: tests/sparseml/modifiers/pruning/constant/test_pytorch.py [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+    assert isinstance(
+        type_, ConstantPruningModifierPyTorch
+    ), "PyTorch ConstantPruningModifier not registered"
+
+
+def _setup_modifier_factory():
+    from sparseml.core.factory import ModifierFactory
+
+    ModifierFactory.refresh()
+    assert ModifierFactory._loaded, "ModifierFactory not loaded"
```
Comment (Satrat) [inline tests/sparseml/modifiers/pruning/constant/test_pytorch.py:173]: This seems like it would be a useful helper function outside of this class, maybe we could move it to a general test helpers file

**Review by Satrat — CHANGES_REQUESTED** [review]
Would be good to test the mask is correctly being added and removed(see comments), but otherwise this is looking good!

File: Makefile (hunk 3 above)
Comment (rahul-tuli) [reply Makefile:28]: Updated to use a regex for files inside sparseml/modifiers

File: tests/sparseml/modifiers/pruning/constant/test_pytorch.py (hunk 4 above)
Comment (rahul-tuli) [reply tests/sparseml/modifiers/pruning/constant/test_pytorch.py:127]: Done!

File: tests/sparseml/modifiers/pruning/constant/test_pytorch.py (hunk 5 above)
Comment (rahul-tuli) [reply tests/sparseml/modifiers/pruning/constant/test_pytorch.py:110]: Done!

File: tests/sparseml/modifiers/pruning/constant/test_pytorch.py (hunk 7 above)
Comment (rahul-tuli) [reply tests/sparseml/modifiers/pruning/constant/test_pytorch.py:173]: Agree moving this to a helper!

File: tests/sparseml/modifiers/pruning/constant/test_pytorch.py (hunk 2 above)
Comment (rahul-tuli) [reply tests/sparseml/modifiers/pruning/constant/test_pytorch.py]: Done!, however it will throw an error now if we do `pytest --collect-only`, works with makefile though with the regex change.

File: tests/sparseml/core/recipe/test_recipe.py (hunk 1 above)
Comment (rahul-tuli) [reply tests/sparseml/core/recipe/test_recipe.py:48]: Added as a part of #1774

File: Makefile [hunk 8]
```diff
@@ -25,7 +25,7 @@ ifneq ($(findstring onnx,$(TARGETS)),onnx)
     PYTEST_ARGS := $(PYTEST_ARGS) --ignore tests/sparseml/onnx
 endif
 ifneq ($(findstring pytorch,$(TARGETS)),pytorch)
-    PYTEST_ARGS := $(PYTEST_ARGS) --ignore tests/sparseml/pytorch --ignore tests/sparseml/modifiers
+    PYTEST_ARGS := $(PYTEST_ARGS) --ignore tests/sparseml/pytorch --ignore-glob 'tests/sparseml/modifiers/*pytorch.py'
```
Comment (Satrat) [inline Makefile:28]: I think this is going to clash with how I defined the tests for the quantization modifier. Lets sync up on the testing structure before this gets merged in

File: src/sparseml/modifiers/obcq/utils/sparsegpt.py [hunk 9]
```diff
@@ -18,7 +18,13 @@
...
 import torch
 import torch.nn as nn
-import transformers
+
+
+try:
+    import transformers
+except ImportError as err:
+    transformers = None
+    transformers_err = err
```
Comment (Satrat) [inline src/sparseml/modifiers/obcq/utils/sparsegpt.py:27]: I think this is also going to clash with a change I made, but I like this solution better! In pair programming today lets go through the merge conflicts

File: tests/sparseml/helpers.py [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+    ]
+
+
+def should_skip_pytorch_tests():
+    try:
+        import torch  # noqa: F401
+    except ImportError:
+        return True
+
+    return os.getenv("NM_ML_SKIP_PYTORCH_TESTS", False)
```
Comment (Satrat) [inline tests/sparseml/helpers.py:82]: Why do we need both this and the change to the makefile?

**Review by Satrat — COMMENTED** [review]
This looks good, but I think our branches have diverged a bit of testing layout. Lets figure it out in pair programming today then get this merged

File: tests/sparseml/helpers.py (hunk 10 above)
Comment (rahul-tuli) [reply tests/sparseml/helpers.py:82]: The glob in makefile only ignores the modifiers/*pytorch.py files; this was done because a few tests in other files were being picked up in base tests, like a few in test_state.py

File: Makefile (hunk 8 above)
Comment (rahul-tuli) [reply Makefile:28]: Updated!

File: src/sparseml/modifiers/obcq/utils/sparsegpt.py (hunk 9 above)
Comment (rahul-tuli) [reply src/sparseml/modifiers/obcq/utils/sparsegpt.py:27]: updated!

**Review by bfineran — DISMISSED** [review]

**Comment by rahul-tuli** [comment]
The failing test `tests/sparseml/pytorch/modifiers/quantization/test_pytorch.py` is fixed by https://github.com/neuralmagic/sparseml/pull/1790

**Review by Satrat — APPROVED** [review]
Looks great pending the session reset PR getting merged!

**Review by bfineran — APPROVED** [review]
